agent/agent.go

```diff
@@ -565,3 +573,22 @@ func formatReportLine(cells ...string) string {
 
 	return fmt.Sprintf(format, strValues...)
 }
+
+// setAgentRedactions mutates an Agent, setting agent-level redactions
+func setAgentRedactions(a *Agent) {
```

Comment:
Few comments here that would be blocking on a non-draft:

- Redactions failing for any reason should block operation of agent - we don't want to bundle up unredacted content. We ought to return any errors, not handle them w/ logging yet.
- Go convention would mean using a method with a pointer receiver here, rather than taking the pkg struct pointer as a first arg.
- We can take this design a step further and make this a referentially transparent function like:

```
package redact

// Flatten takes any number of slices of redacts and returns one slice containing all redacts in argument order
func Flatten(redacts ...[]*redact.Redact) []*redact.Redact { ... }
```

called as:

```
// NOTE: Would this work in agent.NewAgent()? we may not need any mutation
func (a *Agent) Setup() error {
...
hclRedacts, err := hcl.MapRedacts(a.Config.HCL.Agent.Redactions)
... (fail setup on err)
defaultRedacts, err := defaultRedacts()
... (fail setup on err)
redacts := redact.Flatten(hclRedacts, agentDefaults)
a.Redactions = redacts
// Note: double check the order on these i hope i did not get it backwards.
...
}

I think this will do a good job separating the "create a flat slice of redactions from two sources in a guaranteed order" concerns from the problems of sourcing multiple agent-specific redacts and storing those redacts on the agent.

It may even be possible to assign the redactions in agent.NewAgent without any mutation taking place!

Reply:
This is a great idea, thank you! I like the Flatten semantics you've suggested -- will implement that.

Also happy to do it in NewAgent(); that's actually how I originally implemented it: https://github.com/hashicorp/hcdiag/compare/main...dcohen/redaction-exploration . I guess I got nervous about changing one constructor and making it potentially different from the others, but I agree that it's cleaner. Consider it done.

Reply:
Something that hadn't occurred to me even when I was refactoring a.Setup() that we may not even need the step. It's more or less doing part of the product & runner reification work that should be done in `agent.New()` and is left over from when the agent involved a lot more mutation throughout its runtime. The whole system gets a lot simpler if all of the agent data is truly fully initialized before it runs. And in the event we want to take in more config and runners during a.Run() we can offer a better API that's referentially transparent and can be unit tested.

Definitely not in scope for the redaction flow, but if you ever want to take on that refactor in the future it'll be an interesting one!